### PR TITLE
Add gecoding v6 service

### PR DESCRIPTION
**Pull request by VladislavMedved**
## Description

Adds geocoding v6 service that supports forward, reverse and structured input modes

### CODE REVIEW

File: services/geocoding-v6.js [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    limit: v.number,
+    worldview: v.string,
+    autocomplete: v.boolean,
+
+    // structured input fields
+    address_number: v.string,
+    street: v.string,
+    place: v.string,
+    region: v.string,
+    country: v.string,
```
Comment (mpothier) [inline services/geocoding-v6.js:127]: `country` and `countries` seem somewhat redundant. Do we need them both for different modes, or can they be consolidated?

File: services/geocoding-v6.js [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+    countries: v.arrayOf(v.string),
+    proximity: v.oneOf(v.coordinates, 'ip'),
+    types: v.arrayOf(v.oneOf(featureTypes)),
+    bbox: v.arrayOf(v.number),
+    language: v.string,
+    limit: v.number,
+    worldview: v.string,
+    autocomplete: v.boolean,
+
+    // structured input fields
```
Comment (mpothier) [inline services/geocoding-v6.js:122]: It looks like this is missing `address_line1`

File: services/geocoding-v6.js [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+/**
+ * Geocoding API service.
+ *
+ * Learn more about this service and its responses in
+ * [the HTTP service documentation](https://docs.mapbox.com/api/search/geocoding-v6/).
+ */
+var GeocodingV6 = {};
+
+var featureTypes = [
```
Comment (mpothier) [inline services/geocoding-v6.js:17]: Confirm if we have `street` as a feature type or not

File: services/geocoding-v6.js [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+    address_number: v.string,
+    street: v.string,
+    place: v.string,
+    region: v.string,
+    country: v.string,
+    postcode: v.string,
+    locality: v.string
+  })(config);
+
+  config.mode = config.mode || 'normal';
```
Comment (mpothier) [inline services/geocoding-v6.js:132]: I'd propose using a different word for 'normal'. Some ideas are `'simple'`, '`standard'`

File: services/geocoding-v6.js [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+ *   .then(response => {
+ *     const match = response.body;
+ *   });
+ */
+GeocodingV6.forwardGeocode = function(config) {
+  v.assertShape({
+    query: v.required(v.string),
+    mode: v.oneOf('normal', 'structured'),
+    countries: v.arrayOf(v.string),
+    proximity: v.oneOf(v.coordinates, 'ip'),
```
Comment (mpothier) [inline services/geocoding-v6.js:114]: Can you check what object-type is referred to by `v.coordinates`? i.e. does it expect a number array, or does it accept an object with `latitude` and `longitude` properties?

**Review by mpothier — CHANGES_REQUESTED** [review]

File: services/geocoding-v6.js (hunk 3 above)
Comment (VladislavMedved) [reply services/geocoding-v6.js:17]: @krydima confirmed we have `street` feature type

File: services/geocoding-v6.js (hunk 5 above)
Comment (VladislavMedved) [reply services/geocoding-v6.js:114]: it's an array of numbers `v.tuple(v.number, v.number)`

File: services/geocoding-v6.js (hunk 1 above)
Comment (VladislavMedved) [reply services/geocoding-v6.js:127]: I don't think we need it both, we can try to consolidate it to one `countries` parameter

Comment (VladislavMedved) [reply services/geocoding-v6.js]: I've also added `block` and `neighborhood` from v6 api documentation

File: services/geocoding-v6.js [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+ *   bbox: [2.14, 48.72, 2.55, 48.96]
+ * })
+ *   .send()
+ *   .then(response => {
+ *     const match = response.body;
+ *   });
+ */
+GeocodingV6.forwardGeocode = function(config) {
+  v.assertShape({
+    query: v.required(v.string),
```
Comment (mpothier) [inline services/geocoding-v6.js:115]: @VladislavMedved will `query` be required even if intending to use structured mode? Will it throw a validation error if `query` is not included?

**Review by mpothier — APPROVED** [review]
@VladislavMedved this looks good, with my only question about whether an empty string for `query` is required to pass validation when using structured mode.

File: services/geocoding-v6.js (hunk 6 above)
Comment (VladislavMedved) [reply services/geocoding-v6.js]: Good catch! I've made it required only for `standard` mode

File: services/geocoding-v6.js [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+ *     const match = response.body;
+ *   });
+ */
+GeocodingV6.reverseGeocode = function(config) {
+  v.assertShape({
+    longitude: v.required(v.number),
+    latitude: v.required(v.number),
+    countries: v.arrayOf(v.string),
+    types: v.arrayOf(v.oneOf(featureTypes)),
+    bbox: v.arrayOf(v.number),
```
Comment (mpothier) [inline services/geocoding-v6.js:215]: Please double check if `bbox` applies for reverse queries.

**Pull request merged**